src/lib/mac/kmac/kmac.cpp

```diff
@@ -0,0 +1,96 @@
+/*
+* KMAC
+* (C) 2023 Jack Lloyd
+* (C) 2023 Falko Strenzke
+* (C) 2023 René Meusel - Rohde & Schwarz Cybersecurity
+*
+* Botan is released under the Simplified BSD License (see license.txt)
+*/
+
+#include <botan/internal/kmac.h>
+
+#include <botan/internal/cshake_xof.h>
+#include <botan/internal/fmt.h>
+#include <botan/internal/keccak_helpers.h>
+
+namespace Botan {
+
+KMAC::~KMAC() = default;
+
+void KMAC::clear() {
+   zap(m_key);
+   m_cshake->clear();
+}
+
+size_t KMAC::output_length() const {
+   return m_output_bit_length / 8;
+}
+
+Key_Length_Specification KMAC::key_spec() const {
+   // KMAC supports key lengths from zero up to 2²⁰⁴⁰ (2^(2040)) bits:
+   // https://nvlpubs.nist.gov/nistpubs/specialpublications/nist.sp.800-185.pdf#page=28
+   //
+   // However, we restrict the key length to 64 bytes in order to avoid allocation of overly
+   // large memory stretches when client code works with the maximal key length.
+   return Key_Length_Specification(0, 64);
```

Comment:
64 bytes is arbitrary but also 512 bits seems more than sufficient as a MAC key

HMAC allows 0..4095 because in PBKDF2 the HMAC "key" is the passphrase and we don't want to artificially limit the length of passphrases that people can use. This was also required for the TLS 1.2 PRF when using eg DH with large groups since there the shared secret will be the size of the field elements and TLS PRF uses the shared secret as an HMAC key.